### PR TITLE
ORCA-369: As an operator, I would like to create a downloadable script for creating s3 storage buckets.

### PR DESCRIPTION
**Pull request by zroehrich**
## Summary of Changes

Addresses [ORCA-369: As an operator, I would like to create a downloadable script for creating s3 storage buckets.](https://bugs.earthdata.nasa.gov/browse/ORCA-369)

## Changes

* Added DR bucket Terraform template and updated S3 deployment documentation.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Successfully deployed DR S3 buckets in the DR account.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the necessary documentation (remove those that do not apply)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
    - [x] Manual tests (outlined above)
- [x] My code has passed security scanning
    - [x] git-secrets
    - [x] Snyk

### CODE REVIEW

**Review by rizbihassan — CHANGES_REQUESTED** [review]
Found some minor errors and one way to automate the region.

File: modules/dr_buckets/dr_buckets.tf [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+variable "aws_profile" {
+  type    = string
+  description = "AWS CLI profile to use for installation and monitoring of AWS objects."
+}
+
+variable "aws_region" {
+  type        = string
+  description = "AWS Region to create resources in."
+  default = "us-west-2"
+}
```
[the remark this answers is not in the document]
Comment (zroehrich) [reply modules/dr_buckets/dr_buckets.tf:27]: I tried using this and get the following error: `Error: Cycle: data.aws_region.current, provider["registry.terraform.io/hashicorp/aws"]`

[the remark this answers is not in the document]
Comment (rizbihassan) [reply modules/dr_buckets/dr_buckets.tf:27]: Here is the template that worked for me

```# ORCA DR S3 Variables
variable "prefix" {
  type        = string
  description = "Prefix used to prepend to all object names and tags."
}

variable "tags" {
  type        = map(string)
  description = "Tags to be applied to resources that support tags."
  default     = {}
}

variable "cumulus_account_id" {
  type = number
  description = "Account ID of the Cumulus Account"
}

## Terraform Configuration
terraform {
  required_version = ">= 1.5"
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 5.0.0"
    }
  }
}

# Get DR Account ID
data "aws_caller_identity" "current" {}
# Get AWS Region
data "aws_region" "current_region" {}

# ORCA Archive Bucket
resource "aws_s3_bucket" "orca_archive_bucket" {
  bucket = "${var.prefix}-orca-archive"
  tags = var.tags
}

# ORCA Archive Bucket Versioning
resource "aws_s3_bucket_versioning" "orca_archive_versioning" {
  bucket = aws_s3_bucket.orca_archive_bucket.id
  versioning_configuration {
    status = "Enabled"
  }
}

# ORCA Archive Policy
resource "aws_s3_bucket_policy" "orca_archive_cross_account_access" {
  bucket = aws_s3_bucket.orca_archive_bucket.id
  policy = data.aws_iam_policy_document.orca_archive_cross_account_access_policy.json
}

data "aws_iam_policy_document" "orca_archive_cross_account_access_policy" {
  statement {
    sid = "CrossAccPolicyDoc"
    principals {
      type        = "AWS"
      identifiers = [var.cumulus_account_id]
    }

    actions = [
      "s3:GetObject*",
      "s3:RestoreObject",
      "s3:GetBucket*",
      "s3:ListBucket",
      "s3:PutBucketNotification",
      "s3:GetInventoryConfiguration",
      "s3:PutInventoryConfiguration",
      "s3:ListBucketVersions"
    ]

    resources = [
      aws_s3_bucket.orca_archive_bucket.arn,
      "${aws_s3_bucket.orca_archive_bucket.arn}/*",
    ]
  }
  statement {
    sid = "Cross Account Write Access"
    principals {
      type        = "AWS"
      identifiers = [var.cumulus_account_id]
    }

    actions = [
      "s3:PutObject*"
    ]

    resources = [
      "${aws_s3_bucket.orca_archive_bucket.arn}/*"
    ]

    condition {
      test = "StringNotEquals"
      variable = "s3:x-amz-storage-class"
      values = [
        "GLACIER",
        "DEEP_ARCHIVE"
      ]
    }
    condition {
      test = "StringNotEquals"
      variable = "s3:x-amz-acl"
      values = [
        "bucket-owner-full-control"
      ]
    }
  }
}

# ORCA Archive Worm Bucket
resource "aws_s3_bucket" "orca_archive_worm_bucket" {
  bucket = "${var.prefix}-orca-archive-worm"
  tags = var.tags
}

# ORCA Archive Worm Bucket Versioning
resource "aws_s3_bucket_versioning" "orca_archive_worm_versioning" {
  bucket = aws_s3_bucket.orca_archive_worm_bucket.id

  versioning_configuration {
    status = "Enabled"
  }
}

# ORCA Archive Worm Bucket Object Lock
resource "aws_s3_bucket_object_lock_configuration" "orca_achive_worm_object_lock" {
  bucket = aws_s3_bucket.orca_archive_worm_bucket.id
  depends_on = [aws_s3_bucket_versioning.orca_archive_worm_versioning]

  rule {
    default_retention {
      mode = "GOVERNANCE"
      days = 5
    }
  }
}

# ORCA Archive Worm Bucket Policy
resource "aws_s3_bucket_policy" "orca_archive_worm_cross_account_access" {
  bucket = aws_s3_bucket.orca_archive_worm_bucket.id
  policy = data.aws_iam_policy_document.orca_archive_worm_cross_account_access_policy.json
}

data "aws_iam_policy_document" "orca_archive_worm_cross_account_access_policy" {
  statement {
    sid = "Cross Account Access"
    principals {
      type        = "AWS"
      identifiers = [var.cumulus_account_id]
    }

    actions = [
      "s3:PutObject"
    ]

    resources = [
      "${aws_s3_bucket.orca_archive_worm_bucket.arn}/*"
    ]
    condition {
      test = "StringEquals"
      variable = "s3:x-amz-storage-class"
      values = [
        "GLACIER",
        "DEEP_ARCHIVE"
      ]
    }
  }
}

# ORCA Reports Bucket
resource "aws_s3_bucket" "orca_reports_bucket" {
  bucket = "${var.prefix}-orca-reports"
  tags = var.tags
}

# ORCA Reports Bucket Lifecycle Rule
resource "aws_s3_bucket_lifecycle_configuration" "orca_reports_bucket_lifecycle" {
  bucket = aws_s3_bucket.orca_reports_bucket.id

  rule {
    id = "30-day-lifecycle-reports"
    expiration {
      days = 30
    }
    status = "Enabled"
  }
}

# ORCA Reports Bucket Policy
resource "aws_s3_bucket_policy" "orca_reports_cross_account_access" {
  bucket = aws_s3_bucket.orca_reports_bucket.id
  policy = data.aws_iam_policy_document.orca_reports_cross_account_access_policy.json
}

data "aws_iam_policy_document" "orca_reports_cross_account_access_policy" {
  statement {
    sid = "Reports Cross Account Access"
    principals {
      type        = "AWS"
      identifiers = [var.cumulus_account_id]
    }

    actions = [
      "s3:GetObject*",
      "s3:GetBucket*",
      "s3:ListBucket",
      "s3:PutObject",
      "s3:PutObjectAcl",
      "s3:PutBucketNotification"
    ]

    resources = [
      aws_s3_bucket.orca_reports_bucket.arn,
      "${aws_s3_bucket.orca_reports_bucket.arn}/*",
    ]
  }
  statement {
    sid = "Inventory ORCA Reports"
    principals {
      type        = "Service"
      identifiers = ["s3.amazonaws.com"]
    }

    actions = [
      "s3:PutObject*"
    ]

    resources = [
      "${aws_s3_bucket.orca_reports_bucket.arn}/*"
    ]

    condition {
      test = "StringEquals"
      variable = "s3:x-amz-acl"
      values = [
        "bucket-owner-full-control"
      ]
    }
    condition {
      test = "StringEquals"
      variable = "aws:SourceAccount"
      values = [
        "${data.aws_caller_identity.current.account_id}"
      ]
    }
    condition {
      test = "ArnLike"
      variable = "aws:SourceArn"
      values = [aws_s3_bucket.orca_reports_bucket.arn]
    }
  }
 
}

# DR Bucket Outputs
output "orca_achive_bucket_arn" {
  value = aws_s3_bucket.orca_archive_bucket.arn
}

output "orca_reports_bucket_arn" {
  value = aws_s3_bucket.orca_reports_bucket.arn
}

output "orca_achive_worm_bucket_arn" {
  value = aws_s3_bucket.orca_archive_worm_bucket.arn
}

```

File: modules/dr_buckets/dr_buckets.tf [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+variable "cumulus_account_id" {
+  type = number
+  description = "Account ID of the Cumulus Account"
+}
+
+variable "aws_profile" {
+  type    = string
+  description = "AWS CLI profile to use for installation and monitoring of AWS objects."
+}
```
Comment (rizbihassan) [inline modules/dr_buckets/dr_buckets.tf:20]: This can be removed.

File: modules/dr_buckets/dr_buckets.tf [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  }
+}
+
+provider "aws" {
+  profile = var.aws_profile
+
+  ignore_tags {
+    key_prefixes = ["gsfc-ngap"]
+  }
+}
```
Comment (rizbihassan) [inline modules/dr_buckets/dr_buckets.tf:39]: This can be removed since terrform conf above already has the `required_providers`

**Review by rizbihassan — CHANGES_REQUESTED** [review]
Some cleanup.

**Review by rizbihassan — APPROVED** [review]

**Review by rizbihassan — APPROVED** [review]

**Pull request merged**